core/chains/evm/logpoller/log_poller.go

```diff
@@ -1191,17 +1244,300 @@ func (lp *logPoller) fillRemainingBlocksFromRPC(
+				if reqsForAddress, ok := lp.cachedReqsByAddress[addr]; !ok {
+					newReq = NewGetLogsReq(filter)
+					lp.cachedReqsByEventsTopicsKey[eventsTopicsKey] = newReq
+					lp.cachedReqsByAddress[addr] = []*GetLogsBatchElem{newReq}
+				} else {
+					newTopics := [][]common.Hash{filter.EventSigs, filter.Topic2, filter.Topic3, filter.Topic4}
+
+					for i, req := range reqsForAddress {
+						topics := req.Topics()
+						if isTopicsSubset(newTopics, topics) {
+							// Already covered by existing req
+							break
+						} else if isTopicsSubset(topics, newTopics) {
+							// Replace existing req by new req which includes it
+							reqsForAddress[i] = NewGetLogsReq(filter)
+							lp.cachedReqsByAddress[addr] = reqsForAddress
+							break
+						}
+						// Nothing similar enough found for this address, add a new req
+						lp.cachedReqsByEventsTopicsKey[eventsTopicsKey] = NewGetLogsReq(filter)
+						lp.cachedReqsByAddress[addr] = append(reqsForAddress, lp.cachedReqsByEventsTopicsKey[eventsTopicsKey])
+					}
+				}
+			}
+		}
+		lp.newFilters = make(map[string]struct{})
+	}
+	lp.filtersMu.Unlock()
+
+	// Fill fromBlock, toBlock, & blockHash while copying cached reqs into a result array
+	reqs := make([]rpc.BatchElem, 0, len(lp.cachedReqsByEventsTopicsKey))
+	for _, req := range lp.cachedReqsByEventsTopicsKey {
+		reqs = append(reqs, rpc.BatchElem{
+			Method: req.Method,
+			Args: []interface{}{
+				map[string]interface{}{
+					"addresses": req.Addresses(),
+					"topics":    req.Topics(),
+					"fromBlock": fromBlock,
+					"toBlock":   toBlock,
+					"blockHash": blockHash,
+				},
+			},
+			Result: new([]types.Log),
+		})
+	}
+
+	return reqs
+}
+
+// batchFetchLogs fetches logs for either a single block by block hash, or by block range,
+// rebuilding the cached reqs if necessary, sending them to the rpc server, and parsing the results
+// Requests for different filters are sent in parallel batches. For block range requests, the
+// block range is also broken up into serial batches
+func (lp *logPoller) batchFetchLogs(ctx context.Context, fromBlock *big.Int, toBlock *big.Int, blockHash *common.Hash) ([]types.Log, error) {
+	reqs := lp.ethGetLogsReqs(fromBlock, toBlock, blockHash)
+
+	lp.lggr.Debugf("batchFetchLogs: rpcBatchSize=%d, len(reqs)=%d, reqs=%v", lp.rpcBatchSize, len(reqs), reqs)
+	if err := lp.sendBatchedRequests(ctx, lp.rpcBatchSize, reqs); err != nil {
+		return nil, err
+	}
+
+	var logs []types.Log
+	for _, req := range reqs {
+		lp.lggr.Debugf("processing result for req...")
+		if req.Error != nil {
+			return nil, req.Error
+		}
+		res, ok := req.Result.(*[]types.Log)
+		if !ok {
+			return nil, fmt.Errorf("expected result type %T from eth_getLogs request, got %T", res, req.Result)
+		}
+		lp.lggr.Debugf("appending %d logs to result", len(*res))
+		logs = append(logs, *res...)
+	}
+	return logs, nil
+}
+
+func (lp *logPoller) sendBatchedRequests(ctx context.Context, batchSize int64, reqs []rpc.BatchElem) error {
```

Comment:
Hmm... it could, but I'm not sure that would be an improvement.  The reason it breaks it up into batches limited by `batchSize` (default 250) is presumably because that's the maximum number of requests the rpc server can handle at once in parallel. If it has to process each batch in series anyway, maybe sending them all at once would just overwhelm it... not entirely sure but I think I prefer to keep it as-is since that's what we've been doing for blocks. Could be something to experiment with at some point.